caddyfile_test.go

```diff
@@ -0,0 +1,104 @@
+package dynamicdns
+
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/caddyserver/caddy/v2/caddyconfig/caddyfile"
+	"github.com/caddyserver/caddy/v2/caddyconfig/httpcaddyfile"
+	"github.com/google/go-cmp/cmp"
+)
+
+func Test_ParseApp(t *testing.T) {
+	tests := []struct {
+		name    string
+		d       *caddyfile.Dispenser
+		want    string
+		wantErr bool
+	}{
+		{
+			name: "ip_source: upnp",
+			d: caddyfile.NewTestDispenser(`
+			dynamic_dns {
+				ip_source upnp
+			}`),
+			want: ` {
+				"ip_sources": [
+					{
+						"source": "upnp"
+					}
+				]
+			}`,
+		},
+		{
+			name: "ip_source: simple http endpoints",
+			d: caddyfile.NewTestDispenser(`
+			dynamic_dns {
+				ip_source simple_http endpoint http://1.com
+				ip_source simple_http endpoint http://2.com
+			}`),
+			want: ` {
+				"ip_sources": [
+					{
+						"source": "simple_http",
+						"endpoints": ["http://1.com"]
+					},
+                    {
```

Comment:
This line has spaces instead of tabs
```suggestion
					{
```